benchmark/src/submit_pattern.rs

```diff
@@ -0,0 +1,212 @@
+#[derive(Debug, StructOpt, Clone, Copy)]
+pub struct ConstRate {
+    /// Constant number of TXN requests submitted per second.
+    #[structopt(short = "c", long = "constant", default_value = "10")]
+    pub constant: usize,
+    /// Note about internal fields: structopt at master supports '#[structopt(skip)]`
+    /// but not in its latest release version yet. TODO: use `#[structopt(skip)]` when released.
+    /// Submission interval in milliseconds. Used internally, not to be set by user.
+    #[structopt(short = "i", long = "interval_ms", default_value = "1000")]
+    pub interval_ms: usize,
+}
+
+#[derive(Debug, StructOpt, Clone, Copy)]
+pub struct LinearRate {
+    /// Linear function's slope, e.g., increase X TXNs per second.
+    #[structopt(short = "a", long = "slope", default_value = "1")]
+    pub slope: usize,
+    /// Linear function's intercept, e.g., start with X TXNs per second. Must be integer.
+    #[structopt(short = "b", long = "intercept", default_value = "0")]
+    pub intercept: usize,
+    /// Linear function's initial index value. Used internally, not to be set by user.
+    #[structopt(short = "i", long = "interval_ms", default_value = "0")]
+    pub idx: usize,
+    /// Submission interval in milliseconds. Used internally, not to be set by user.
+    #[structopt(short = "i", long = "interval_ms", default_value = "1000")]
+    pub interval_ms: usize,
+}
+
+/// Determine what is the best submission interval (in milliseconds) for given number of
+/// submissions. Here "best" means we submit 1 <= x < 10 requests in an interval.
+/// For example, for 100 req / s, return (1 req, 10 ms).
+fn adjust_submission_interval_ms(num_reqs_: usize, interval_ms_: usize) -> (usize, usize) {
+    let (mut num_reqs, mut interval_ms) = (num_reqs_, interval_ms_);
+    while num_reqs > 10 && interval_ms > 10 {
+        num_reqs /= 10;
+        interval_ms /= 10;
+        info!(
+            "Submit pattern is adjusted to {} reqs per {} ms",
+            num_reqs, interval_ms
+        );
+    }
+    (num_reqs, interval_ms)
+}
+
+/// Defines how a pattern should return the amount of heading items from a vector.
+pub trait ItemsToConsume {
+    /// Give implementor a chance to config itself.
+    fn init(&mut self);
+    /// Move a certain amount of items from given vector.
+    fn get_next<T>(&mut self, vec: &mut Vec<T>) -> Vec<T>;
+}
+
+impl ItemsToConsume for FloodAll {
+    fn init(&mut self) {}
+
+    fn get_next<T>(&mut self, vec: &mut Vec<T>) -> Vec<T> {
+        vec.drain(0..vec.len()).collect()
+    }
+}
+
+impl ItemsToConsume for ConstRate {
+    fn init(&mut self) {
+        let (amount, interval_ms) = adjust_submission_interval_ms(self.constant, 1000);
+        self.constant = amount;
+        self.interval_ms = interval_ms;
+    }
+
+    fn get_next<T>(&mut self, vec: &mut Vec<T>) -> Vec<T> {
+        let n = std::cmp::min(self.constant, vec.len());
+        vec.drain(0..n).collect()
+    }
+}
+
+impl ItemsToConsume for LinearRate {
```

Comment:
Aren't all of `FloodAll`, `ConstRate` and `LinearRate` are just `LinearRate` with different parameters?
Can you squash all of them, then you don't need to re-implement ItemsToConsume (4 times, with many matches, that lose the point of implementing the trait)